frontend/server/src/main/java/org/pytorch/serve/wlm/WorkerThread.java

```diff
@@ -481,16 +484,29 @@ public void setState(WorkerState newState, int status) {
         listener.notifyChangeState(
                 model.getModelVersionName().getVersionedModelName(), newState, status);
         logger.debug("{} State change {} -> {}", getWorkerName(), state, newState);
-        long timeTaken = System.currentTimeMillis() - startTime;
+        long currentTS = System.currentTimeMillis();
+        long timeTaken = currentTS - startTime;
         if (state != WorkerState.WORKER_SCALED_DOWN) {
             // Don't update the state if it was terminated on purpose.. Scaling in..
             this.state = newState;
         }
+
         if (state == WorkerState.WORKER_MODEL_LOADED) {
             workerLoadTime.setValue(String.valueOf(timeTaken));
             workerLoadTime.setTimestamp(
                     String.valueOf(TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())));
             loggerTsMetrics.info("{}", workerLoadTime);
+            if (recoveryStartTS > 0) {
+                logger.info("Auto recovery succeeded, reset recoveryStartTS");
+                recoveryStartTS = 0;
+            }
+        } else if (state == WorkerState.WORKER_STOPPED) {
+            if (recoveryStartTS == 0) {
+                recoveryStartTS = currentTS;
```

Comment:
Is `recoveryStartTS = currentTS;` only valid in case of `WorkerState.WORKER_STOPPED`

What about `WorkerState.WORKER_SCALED_DOWN`? 

Also, does the current logic handle the case when the thread is dying because of OOM

Reply:
WORKER_SCALED_DOWN is used in model unregistration or scale down request. it does not trigger backend worker retry.

Any exception such as OOM will trigger worker stage changed to WorkerState.WORKER_STOPPED and then retry. That's why recoveryStartTS = currentTS only happen on WorkerState.WORKER_STOPPED